### PR TITLE
Build Codecs with macros

### LINKED ISSUE #26: Use macros, slim down fat
`impls.rs` could be cleaned up with macros, along with some other minor changes here and there around the repo.

It could also get the same treatment as https://github.com/multiformats/rust-multihash where each base code is a feature, which may be useful for those looking to use `no_std` to slim down crate size although that might be a bit overkill.

I'd be interested in making these changes but figured opening an issue first before making a large PR would be better than not, especially if people are interested in helping out.

### PR DESCRIPTION
**Pull request by mriise**
Just might mess up the documentation a bit...
I didn't jump ahead and add documentation as part of the macro because it started to get a bit more complicated than I was hoping. If that level of documentation is still desired, I don't mind adding it in.
fixes #26

### CODE REVIEW

**Comment by mriise** [comment]
decided to go ahead and do it anyway.

File: src/impls.rs [hunk 1]
```diff
@@ -1,6 +1,52 @@
...
+
+            fn decode<I: AsRef<str>>(input: I) -> Result<Vec<u8>> {
+                Ok($encoding.decode(input.as_ref().as_bytes())?)
+            }
+        }
+    };
+    (#[$doc:meta] $type:ident, $encoding:expr; $(#[$doc2:meta] $type2:ident, $encoding2:expr;)+) => {
+        derrive_base_encoding! (#[$doc] $type, $encoding;);
+        derrive_base_encoding! ($(#[$doc2] $type2, $encoding2;)+);
+    };
```
Comment (vmx) [inline src/impls.rs:24]: That's not needed. You can change the rule at the top.

Wrap it in a `($(…)*)`. Then wrap the whole contents (the stuff between the curly brackets) in `$(…)*`.

So it reads line:

```rust
($(#[$doc:meta] $type:ident, $encoding:expr;)*) => {
    $(
…
    )*
};
```

I also had to look that up. https://medium.com/@phoomparin/a-beginners-guide-to-rust-macros-5c75594498f1 helped a lot.

**Review by vmx — COMMENTED** [review]
Thanks, that looks good, also splitting it into two macros and not trying to make it work within a single one. See inline comments for a few minor things.

**Review by vmx — COMMENTED** [review]
Thanks, it looks good now and indeed way cleaner/more maintainable then before.

You've based your code on an old version without the Base36 encoding. Could you please rebase this PR? I'd expect that an interactive rebase or merge would be more complicated then just starting from current HEAD and applying those changes again. You could then force push the result to this PR.

**Comment by vmx** [comment]
Sorry that you have to rebase again, due to the merge of #29. But at least afterwards the CI should be green :)

**Comment by mriise** [comment]
seems case insensitivity tests for Base36Lower and Upper are failing

**Review by vmx — APPROVED** [review]
Thanks!

**Pull request merged**